ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs

```diff
@@ -54,19 +90,46 @@ type PipeOperatorVisitor(sameLinePipeHints: SameLinePipeHints) =
 type PipeChainHighlightingProcess(logger: ILogger, fsFile, settings: IContextBoundSettingsStore, daemonProcess: IDaemonProcess) =
     inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
 
-    let adornExprs logKey (exprs : (IReferenceExpr * ITreeNode)[]) =
+    let isRightPipeArgApplicable (expr: ITreeNode) (pipeResultType: FSharpType) =
+        if not (isUnit pipeResultType) then true else
+
+        let rec skipFunArgs (t: FSharpType) =
+            if t.IsFunctionType then skipFunArgs t.GenericArguments[1] else t
+
+        let invokedRefExpr =
+            match expr.As<IFSharpExpression>().IgnoreInnerParens() with
+            | :? IReferenceExpr as refExpr ->
+                refExpr
+            | :? IPrefixAppExpr as expr ->
+                expr.InvokedReferenceExpression
+            | _ -> null
+
+        if isNull invokedRefExpr then true else
+        let returnType = invokedRefExpr.Reference.GetFcsSymbol() |> getReturnType
+        match returnType with
+        | Some t ->
+            skipFunArgs t
+            |> isUnit
+            |> not
+        | None -> false
+
+    let adornExprs logKey (exprs : (IReferenceExpr * ITreeNode * bool)[]) =
         use _swc = logger.StopwatchCookie(sprintf "Adorning %s expressions" logKey, sprintf "exprCount=%d sourceFile=%s" exprs.Length daemonProcess.SourceFile.Name)
         let highlightingConsumer = FilteringHighlightingConsumer(daemonProcess.SourceFile, fsFile, settings)
 
-        for refExpr, exprToAdorn in exprs do
+        for refExpr, exprToAdorn, isLeft in exprs do
             if daemonProcess.InterruptFlag then raise <| OperationCanceledException()
 
             let symbolUse = refExpr.Reference.GetSymbolUse()
             if isNull symbolUse then () else
 
-            let _, fcsType = symbolUse.GenericArguments[0]
+            let _, fcsType = symbolUse.GenericArguments[if isLeft then 0 else 1]
+            let isApplicable = isLeft || isRightPipeArgApplicable exprToAdorn fcsType
+
+            if not isApplicable then () else
             let range = exprToAdorn.GetNavigationRange().EndOffsetRange()
-            highlightingConsumer.AddHighlighting(TypeHintHighlighting(fcsType.Format(symbolUse.DisplayContext), range))
+            let displayContext = symbolUse.DisplayContext.WithShortTypeNames(true)
```

Comment:
Could be nice to be in a separate PR, but OK. Could you add some tests for this change, please?